### PR TITLE
Cmake improvements

### PR DESCRIPTION
**Pull request by RAMitchell**
-Use cmake provided functions instead of directly modifying flags, where possible
-Group sources for visual studio
-Default to release build
-Refactor to improve readability

Google test to be added soon

### CODE REVIEW

**Comment by RAMitchell** [comment]
@superbobry can you review this please. Note that I have changed cmake to only produce a release configuration unless otherwise specified.

File: CMakeLists.txt [hunk 1]
```diff
@@ -62,99 +33,72 @@ include_directories (
...
-    src/gbm/*.cc
-    src/metric/*.cc
-    src/objective/*.cc
-    src/tree/*.cc
+file(GLOB_RECURSE SOURCES
     src/*.cc
+    src/*.h
 )
 
+# TODO: Create rabit cmakelists.txt
```
Comment (superbobry) [inline CMakeLists.txt]: 👍

File: CMakeLists.txt [hunk 2]
```diff
@@ -62,99 +33,72 @@ include_directories (
...
 
   cuda_add_library(gpuxgboost ${CUDA_SOURCES} STATIC)
   target_link_libraries(gpuxgboost nccl)
   list(APPEND LINK_LIBRARIES gpuxgboost) 
   list(APPEND SOURCES plugin/updater_gpu/src/register_updater_gpu.cc) 
-else()
-  set(CUDA_OBJS "")
 endif()
 
+add_library(rabit STATIC ${RABIT_SOURCES})
```
Comment (superbobry) [inline CMakeLists.txt]: Ideally, we should allow the user to configure the backend, in the same way Makefiles do.

**Comment by ericxu10101** [comment]
I always get error on "include <gtest/gtest.h>". any instructions or document on how to setup gtest for tests/cpp ? I am using mac.

**Pull request merged**